README.md
# @segment/source-stripe

Stripe data source for Segment.

## Requirements

- Node >= 4.0.0

## Quickstart

```
node bin/stripe \
  --secret    [ STRIPE_SECRET ] \
  --write-key [ WRITE_KEY ]
```

## Metrics

The following metrics are collected by this source:

Name                | Type      | Description
----                | ----      | -----------
`http.response.payload_size` | Histogram | HTTP response content length from source API response.
`http.request.duration` | Histogram | How long an HTTP request took, in ms.
`http.request.total` | Counter | Total number of HTTP requests to source API.
`http.response.status_code` | Counter | HTTP response status code from source API response.


HISTORY.md

v2.4.2 / 2017-09-07
===================

  * Don't modify objects received from the downloader

v2.4.1 / 2017-09-07
===================

  * Fetch transactions via transfers in a memory-scalable way (#63)

v2.3.1 / 2017-08-29
===================

  * Migrate to circle 2.0 and debian linux (#61)

v2.2.11 / 2017-08-24
====================

  * Upgraded source-runner

v2.2.10 / 2017-08-24
====================

  * Upgraded source-runner

v2.2.9 / 2017-08-23
===================

  * Upgraded source-runner to v4.17.1

v2.2.8 / 2017-08-14
===================

  * Re-release to deploy from ECR

v2.2.7 / 2017-08-11
===================

  * Pull golang-private from ECR

v2.2.6 / 2017-08-11
===================

  * Publish to ECR/source-stripe

v2.2.5 / 2017-08-09
===================

  * Use workspace/project slugs in metrics

v2.2.4 / 2017-08-09
===================

  * Upgraded source-runner

v2.2.2 / 2017-08-09
===================

  * Added api request usage metrics

v2.2.1 / 2017-08-04
===================

  * Added LSP logging (#62)

v2.1.2 / 2017-08-03
===================

  * Added missing properties to "cards"

v2.1.1 / 2017-08-01
===================

  * Rewrite in Go (#60)

v1.16.3 / 2017-07-14
====================

  * Updated source-runner

v1.16.2 / 2017-06-06
====================

  * Bumped source-runner version

v1.16.1 / 2017-05-31
====================

  * Updated source-runner (nsq retries)

v1.15.4 / 2017-05-22
====================

  * update source-runner 

v1.15.2 / 2017-04-01
====================

  * temporarily reverting source-runner

v1.14.10 / 2017-02-15
=====================

  * fixing lint

v1.14.9 / 2017-02-15
====================

  * fixing config so it uses kit camelCase

v1.14.6 / 2017-02-14
====================

  * adding stripe version flag

1.3.1 / 2015-03-08
==================

- Update to `@segment/source@0.6.0` to add GetContext and SetContext RPC methods
- Update to `node-kit@3.1.0` to fix hang caused by metrics socket

1.3.0 / 2015-03-08
==================

- Switch from commander to node-kit for command-line option parsing
- Start collecting runtime stats (via node-kit)

1.0.12 / 2015-09-25
===================

  * adding discounts to invoices and subscriptions

1.0.11 / 2015-09-22
===================

    ?_?

1.0.10 / 2015-09-21
===================

  * updating segmentio/soruce and adding error logging

1.0.9 / 2015-09-19
==================

  * adding empty date check pattern

1.0.8 / 2015-09-18
==================

  * adding retries to stripe
  * fixing subscriptions.plan.id and removing subscriptions.discount
  * empty plan object on invoice_items

1.0.7 / 2015-09-LOL
==================

  * Update to latest sources
  * Remove history
  * Buile new version
  * Override entrypoint
  * Update Dockerfile
  * Update Readme.md
  * Update circle
  * Update Dockerfile


api/client.go
package api

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"github.com/apex/log"
	"github.com/nu7hatch/gouuid"
	"github.com/pkg/errors"
	"github.com/segmentio/go-source"
	"github.com/segmentio/go-source/source-logger"
	"github.com/segmentio/ur-log"
	"net/http"
	"os"
	"time"
)

const stripeApiVersion = "2016-07-06"

type clientImpl struct {
	httpClient   HttpClient
	baseUrl      string
	secret       string
	throttler    *Throttler
	sourceClient source.Client
	sourceLogger SourceLogger
}

func (c *clientImpl) GetList(ctx context.Context, req *Request) (*ObjectList, error) {
	output := &listResponse{}
	if err := c.get(ctx, req, output); err != nil {
		return nil, err
	}

	result := &ObjectList{
		HasMore: output.HasMore,
	}
	for _, obj := range output.Data {
		result.Objects = append(result.Objects, obj)
	}

	return result, nil
}

func (c *clientImpl) GetObject(ctx context.Context, req *Request) (Object, error) {
	output := Object{}
	if err := c.get(ctx, req, &output); err != nil {
		return nil, err
	}

	return output, nil
}

func (c *clientImpl) prepareRequest(req *Request) (*http.Request, error) {
	url := req.Url
	if url[0] == '/' {
		url = c.baseUrl + url
	}

	httpReq, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	newQs := httpReq.URL.Query()
	for key, value := range req.Qs {
		newQs[key] = value
	}
	httpReq.URL.RawQuery = newQs.Encode()

	httpReq.Header.Set("Stripe-Version", stripeApiVersion)
	for key, value := range req.Headers {
		httpReq.Header[key] = value
	}
	httpReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.secret))

	return httpReq, nil
}

func (c *clientImpl) get(ctx context.Context, req *Request, output interface{}) error {
	httpReq, err := c.prepareRequest(req)
	if err != nil {
		ctx, _ := urlog.GetContextualLogger(ctx, nil, log.Fields{"request": req})
		return urlog.WrapError(ctx, err, "failed to prepare request")
	}

	uv4, err := uuid.NewV4()
	if err != nil {
		return urlog.WrapError(ctx, err, "failed to generate uuid")
	}

	ctx, logger := urlog.GetContextualLogger(ctx, nil, log.Fields{
		"request": log.Fields{
			"id":      uv4.String(),
			"url":     httpReq.URL.String(),
			"headers": httpReq.Header,
		},
	})

	c.throttler.Use()

	ts := time.Now()
	logger.Info("http request")
	metricTags := []string{}
	workspaceSlug := os.Getenv("SEGMENT_WORKSPACE_SLUG")
	projectSlug := os.Getenv("SEGMENT_SOURCE_SLUG")
	if projectSlug != "" && workspaceSlug != "" {
		metricTags = append(metricTags, fmt.Sprintf("project:%s/%s", workspaceSlug, projectSlug))
	}
	c.sourceClient.StatsIncrement("stripe.requests", 1, metricTags)
	resp, err := c.httpClient.Do(httpReq)
	c.sourceLogger.RequestSent(req.LogCollection, httpReq.URL.String(), sourcelogger.Metadata{"uuid": uv4.String()})
	if err != nil {
		return urlog.WrapError(ctx, err, "error performing request")
	}
	defer resp.Body.Close()

	buffer := &bytes.Buffer{}
	if _, err := buffer.ReadFrom(resp.Body); err != nil {
		return urlog.WrapError(ctx, err, "error reading response")
	}

	duration := time.Now().Sub(ts)
	metricTags = append(metricTags,
		fmt.Sprintf("status_code:%d", resp.StatusCode),
		fmt.Sprintf("status_code_bucket:%dxx", resp.StatusCode/100),
	)
	c.sourceClient.StatsIncrement("stripe.responses", 1, metricTags)
	c.sourceClient.StatsHistogram("stripe.response.payload_size", int64(buffer.Len()), metricTags)
	c.sourceClient.StatsHistogram("stripe.response.latency", duration.Nanoseconds()/1000000, metricTags)

	headersBuffer := &bytes.Buffer{}
	resp.Header.Write(headersBuffer)
	logMetadata := sourcelogger.Metadata{
		"uuid":    uv4.String(),
		"status":  resp.Status,
		"headers": headersBuffer.String(),
	}
	c.sourceLogger.ResponseReceived(req.LogCollection, httpReq.URL.String(), logMetadata, duration, buffer.String())

	ctx, logger = urlog.GetContextualLogger(ctx, logger, log.Fields{
		"response": log.Fields{
			"headers": resp.Header,
			"status":  resp.Status,
			"body":    buffer.String(),
		},
	})
	logger.Debug("http response")

	if resp.StatusCode == 200 {
		decoder := json.NewDecoder(bytes.NewReader(buffer.Bytes()))
		decoder.UseNumber()
		if err := decoder.Decode(output); err != nil {
			return urlog.WrapError(ctx, err, "error decoding response")
		}
		return nil
	}

	err = urlog.WrapError(ctx, errors.New("unexpected response status code"), "")
	if resp.StatusCode >= 500 || resp.StatusCode == 429 {
		// transient errors
		return err
	}

	// non-transient errors
	isAuthRelated := resp.StatusCode == 401
	return &permanentError{
		wrappedError:  err.(wrappedError),
		isAuthRelated: isAuthRelated,
	}
}

func NewClient(opts *ClientOptions) Client {
	if opts.BaseUrl == "" {
		opts.BaseUrl = "https://api.stripe.com"
	}

	return &clientImpl{
		httpClient:   opts.HttpClient,
		baseUrl:      opts.BaseUrl,
		secret:       opts.Secret,
		throttler:    NewThrottler(opts.MaxRps, time.Second),
		sourceClient: opts.SourceClient,
		sourceLogger: opts.SourceClient.Log(),
	}
}
